torchdata/stateful_dataloader/incremental_state.py

```diff
@@ -0,0 +1,178 @@
+from typing import Any, Dict, Optional
+
+import torch
+
+_WORKER_ID = "worker_id"
+_FETCHER_STATE = "fetcher_state"
+_FETCHER_ENDED = "fetcher_ended"
+_DATASET_STATE = "dataset_state"
+_DATASET_ITER_STATE = "dataset_iter_state"
+
+
+class Flattener:
+    @classmethod
+    def flatten(cls, nested_data: Optional[Dict[str, Any]], prefix: str = "", separator: str = "/"):
+        if nested_data is None:
+            return None
+        data = {}
+        if isinstance(nested_data, dict):
+            for key, value in nested_data.items():
+                flat = cls.flatten(
+                    value,
+                    prefix=f"{prefix}{separator}{key}" if len(prefix) > 0 else key,
+                    separator=separator,
+                )
+                data.update(flat)
+        else:
+            if len(prefix) > 0:
+                data[prefix] = nested_data
+            else:
+                data = nested_data
+        return data
+
+    @classmethod
+    def unflatten(cls, flat_data: Optional[Dict[str, Any]], separator: str = "/"):
+        if flat_data is None:
+            return None
+        nested_data = {}
+        for key, value in flat_data.items():
+            _keytoks = key.split(separator)
+            prefix = _keytoks[0]
+            suffix = separator.join(_keytoks[1:])
+            if len(_keytoks) == 1:
+                nested_data[prefix] = value
+                continue
+            if prefix not in nested_data:
+                nested_data[prefix] = {}
+            nested_data[prefix][suffix] = value
+        # now go through nested_data and combine them all
+        for k, v in nested_data.items():
+            if isinstance(v, dict):
+                nested_data[k] = cls.unflatten(v, separator=separator)
+        return nested_data
+
+
+class DeletionTombStone:
+    pass
+
+
+class IncrementalState:
+    def __init__(self, initial_state: Optional[Dict[str, Any]]):
+        self.flat_state = Flattener.flatten(initial_state)
+
+    def generate_delta(self, new_state: Dict[str, Any]):
+        new_flat_state = Flattener.flatten(new_state)
+        delta_flat_state = {}
+        all_keys = set()
+        if self.flat_state:
+            all_keys = set(self.flat_state.keys())
+        all_keys = all_keys.union(new_flat_state.keys())
+
+        for key in all_keys:
+            if self.flat_state is None or key not in self.flat_state:
+                # New key, retain it
+                delta_flat_state[key] = new_flat_state[key]
+                continue
+
+            if key not in new_flat_state:
+                # Key deletion, put in a tombstone
+                delta_flat_state[key] = DeletionTombStone()
+                continue
+
+            prev_value, new_value = self.flat_state[key], new_flat_state[key]
+            try:
+                if isinstance(prev_value, torch.Tensor) and isinstance(new_value, torch.Tensor):
+                    if torch.allclose(prev_value, new_value):
```

Comment:
I wonder if we should actually use down-to-the-bit equality here, since we do actually want to pick up any changes, even if noise

Reply:
That's a good point, I will switch to use torch.equal